tests/expose_test.go

```diff
@@ -484,10 +889,64 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 
 				By("Starting a job which tries to reach the VMI via the ClusterIP service.")
 				job = runHelloWorldJob(serviceIP, servicePort, vm.Namespace)
+				jobCleanupFuncs = append(jobCleanupFuncs, cleanupJob(job.GetName(), job.GetNamespace()))
 
 				By("Waiting for the job to report a failed connection attempt.")
 				Expect(tests.WaitForJobToFail(job, 120*time.Second)).To(Succeed())
-			})
+
+				// this way, the test does not attempt to stop the VM in the `AfterEach` section
+				vm = nil
```

Comment:
line 881
```
				By("Deleting the VM.")
				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())					Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
```

1. u dont wait here for the vm to gone,
2. the vmi = nil should be put directly after that, else the assert below it would break,
the AfterEach will try to delete and assert again (but it will be silent and not that critical - just if you want to be more accurate)
3. you can use WaitForVirtualMachineToDisappearWithTimeout in your AfterEach where you "stop" maybe please

Reply:
> > line 881
> > ```
> > 				By("Deleting the VM.")
> > 				Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())					Expect(virtClient.VirtualMachine(vm.Namespace).Delete(vm.Name, &k8smetav1.DeleteOptions{})).To(Succeed())
> > 				tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
> > ```
> > 
> > 
> > 
> > 1. u dont wait here for the vm to gone, (just vmi) - maybe will be good to remove this from the test itself and let the AfterEach do the job always ?
> 
> The goal of this test is to "Verify an exposed service of a VM is not functional after VM deletion". As such, the test must delete the VM.

It does delete the vmi, it doesnt wait for the vm to be gone as well, as you do in the AfterEach.

> 
> > 1. the vmi = nil should be put directly after that, else the assert below it would break,
> >    the AfterEach will try to delete and assert again (but it will be silent and not that critical - just if you want to be more accurate)
> 
> I think I understand what you mean; added that in a defer, so I always reset the pointer, no matter what happens in the test (only after the VM object is deleted).
> 
> > 1. you can use WaitForVirtualMachineToDisappearWithTimeout in your AfterEach where you "stop" maybe please
> 
> That function's input is a VMI, and we only have a VM. 'GET'ing the VMI from the VM in order to re-use that function (when I only want to cleanup) is going too far imo.

`vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})`
as the test does, or whatever you want, just in order that the delete of the test would be equal to the delete of the AfterEach imo

Reply:
Added a when context that hopefully helps making this clearer; now there's a `When` clause for when just start / stop the VM, and another `When` clause when we just delete the VM .

Are you OK with this option ?

Reply:
i dont understand why `When` or semantics is related here to be honestt,
i meant that there is asymmetric between the delete in the test versus the delete in the AfterEach,
the test doesnt wait for the VM itself to gone, where the AfterEach **did** i think ?

**did** - because where is it gone ? is it just me or its gone ? (where you delete the vmi in AfterEach in case of exception as it was before, (and i think you added also wait for vm to be gone).

moreoever i think that code reuse should be done as more as possible, but lets please first discuss where the AfterEach gone

Reply:
The delete on the test did not wait because it was the last test on a sequence of tests; once it finished, the next module would destroy the namespace.

The delete on the `AfterEach` needs to wait because there will be more tests executed afterwards (at least the delete test on ipv6 family), in which a new VM will have to be created.

There are now 2 when clauses:
  - 1 where the VM is created once; the VM is started in the tests - where the service is also exposed. The VM is then stopped in the test teardown. Since it is only stopped, there's no need to wait for the VM.
  - another where the VM is created in the test setup (since the test itself **deletes** the VM after it is gone.

Yes, I stopped waiting for the VM to be gone since it is either re-used, or we create a new one (this corner-case happens only on the last VM test, where it ends up deleting the VM).

Does it make sense now ?

Reply:
imho i wouldnt do this split, it adds confusion, moreover

1. about `When("the VM object must be created")`
in case of expection, the vmi wont be deleted, since it wont reach the delete step, i prefer much the previous state that you did, where there is a guard using the AfterEach, even vmi isnt deleted here

2. `When("the VM object can be re-used")`
i am not sure we want to be depend on the next module (as you said in the first paragraph), and so on, unless its what there was before and considered non network test.
next module might be skipped and so on.

i personally find it confusing and not robust, i think each test should be self contained, at least to the extend it was before.
those assumptions make it harder to maintain

Reply:
OK, I see this has taken a turn for the worse. I'll get this back to it's initial state.

I'll change the test to wait everywhere since this seems to be upsetting you (thus aligning VM removal throughout the VM test).

This is the function I'm planning on using:

```golang
		cleanupVM := func(virtClient kubecli.KubevirtClient, namespace string, vmName string) {
			By("Deleting the VM ...")
			err := virtClient.VirtualMachine(vm.GetNamespace()).Delete(vm.GetName(), &k8smetav1.DeleteOptions{})
			Expect(err).NotTo(HaveOccurred(), "should have been able to delete the VM object")
			Eventually(func() bool {
				_, errVMI := virtClient.VirtualMachineInstance(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
				_, errVM := virtClient.VirtualMachine(vm.GetNamespace()).Get(vm.GetName(), &k8smetav1.GetOptions{})
				return errVMI != nil && errVM != nil && errors.IsNotFound(errVMI) && errors.IsNotFound(errVM)
			}, 120*time.Second, 1*time.Second).Should(BeTrue())
		}
```

Reply:
Thanks, about the snippet, you can use `SatisfyAll(HaveOccurred(), WithTransform(errors.IsNotFound, BeTrue()))`
with two `Eventually` imo
and i guess that vm is always there when using it right ? (because of the first Expect)

Reply:
Imo its more ginkgo oriented and modern than
`return errVMI != nil && errVM != nil && errors.IsNotFound(errVMI) && errors.IsNotFound(errVM)`

also its logical, you want the vmi to be gone, and then you want the vm to be gone,
no need to check again and again the vmi after its done, just because you wait vm to gone (i know it doesnt take a lot of time, still better logic and readability)